lib/elixir/lib/kernel.ex

```diff
@@ -2442,18 +2442,36 @@ defmodule Kernel do
 
   """
   defmacro first .. last do
-    case is_float(first) or is_float(last) or
+    # require Logger
+    # Logger.debug inspect [first, last]
+    case is_tuple(first) and tuple_size(first) == 2 or
+         is_tuple(last) and tuple_size(last) == 2 or
+         is_float(first) or is_float(last) or
          is_atom(first) or is_atom(last) or
-         is_binary(first) or is_binary(last) do
+         is_binary(first) or is_binary(last) or
+         is_list(first) or is_list(last)
+         do
       true ->
-        raise ArgumentError,
-          "ranges (left .. right) expect both sides to be integers, " <>
-          "got: #{Macro.to_string({:.., [], [first, last]})}"
+        range_error(first, last)
       false ->
+        for val <- [first, last] do
+          case val do
```

Comment:
well, this was never meant to cover every possible case, or to replace Range.validate_range!/2
The idea was to avoid assigning ranges, that we know for sure they  are not valid,
looking at it now, i would remove the check for the functions,
i would leave the list as
` :{}, :%{}, :..` and `:-, :+`
my main concern is  not about missing something that shouldn't be a range, but rejecting something that is a valid one (ie. a false negative)